DAQ string parsing documentation + blackbox

anaymehta opened the pull request:
Added documentation for DAQ string parsing in expectation of future adjustments.

anaymehta committed:
Create readme.md

Documented the string parsing methodology used in main.ino and relied upon by main.py

anaymehta committed:
Update main.py

Included reference to string parsing documentation

anaymehta commented:
For future reference, should such minor changes (comments, readmes) warrant pull requests, or would it be better to simply commit them to master directly?

ashvinnagarajan commented:
I like seeing the Pull Request! Its super quick to review so its not an issue. In general, Pull Requests should always be super small and manageable to read. Typically the max for one PR is 100 lines of code!

ashvinnagarajan reviewed (approved):
I'll approve this but we should do some more quick changes to improve this. Instead of having an explicit number of characters, we can have a setup where you parse until you reach a colon or semicolon like "temp:78.2; altd:11; garb:124124.124;"

anaymehta commented:
Specifications for this pull request are to be expanded to include code for added blackbox functionality, headed by @RupinMittal and @JacobSandler

JacobSandler committed:
Updated test script for black box

Included are two files one pushes the data to the firebase using "." characters to parse the data, and sends the data to a seperrateblack box file in case of disconnect. The other file interprets the black box strings and reuploads them to the firebase once connection is reestablished.

anaymehta commented:
@JacobSandler and @RupinMittal have done some amazing work on implementing the blackbox (a local copy of the data) here! Check out the [firebase](https://console.firebase.google.com/u/0/project/smv-daq/database/smv-daq/data) and see if you can tell which trials required the blackbox and which ones didn't ;)

For now we've only altered the test script, and when we're happy with it we'll update the main script too.

abdullahelqaq commented on sensors/main/readme.md:
```diff
@@ -0,0 +1,5 @@
+### Documentation for string parsing.
+main.ino currently prints all received data values in a single string, with each datum spanning 6 contiguous characters (4 for the label and 2 for the value). This allows [main.py](https://github.com/Bruin-SuperMileage/EV/blob/master/daq/panda/main.py) to read the string in 6 character increments, thereby restoring the original data.
+
+##### Areas for improvement
+- There is no guarantee that the received data values will span only 2 characters. Provisions must be made so that all data values are measured to the same number of digits without any losses in precision
```
I really like this. Can we get documentation like this in the `daq/` directory too?

anaymehta replied:
Great idea! We'll definitely add some documentation before we merge.
On another note, I think the "Areas for Improvement" section is outdated now, because we've switched our parsing method to accommodate varying float lengths, so we'll update that too.

abdullahelqaq reviewed (commented):
Looks good! Just want to push for more README documentation :)

anaymehta committed:
Updated sensor readme

Accounted for new semicolon parsing

anaymehta committed:
Create readme.md

Added identical readme to test file as well, not sure if needed

anaymehta committed:
create readme.md

brief documentation for testScript and blackbox

anaymehta commented:
Alright, final step should be to update main.py to include the blackbox code, and then we should be good to merge!

abdullahelqaq reviewed (approved):
Looks good!

JacobSandler committed:
Changed String Parsing to use ;

anaymehta commented on daq/panda/test/interpreter.py:
```diff
@@ -0,0 +1,108 @@
+
+        numValues = 22 #expected number of data inputs
+        data = line.split(".") #splits each data point into a list 
+        timeName = data[22][: -1]
```
should probably change this to data[numValues]

anaymehta replied:
or data[data.size()] (see below)

anaymehta commented on daq/panda/test/interpreter.py:
```diff
@@ -0,0 +1,108 @@
+        #now = datetime.now() #set time
+        
+
+        numValues = 22 #expected number of data inputs
```
would we perhaps be better off setting numValues to data.size()? That way in the case that one of the sensors fails the code will continue to work, avoiding indexing issues with the array

anaymehta commented on daq/panda/test/interpreter.py:
```diff
@@ -0,0 +1,108 @@
+        timeName = data[22][: -1]
+
+        for i in range (0, numValues):
+                inputs_dict[prefix[i]] = float(data[i]) #may fail on first run, simply re-run
```
i feel like we might want to leave in the labels, again in the case that a sensor fails and isn't able to provide its data, so that we don't get misaligned data

anaymehta committed:
updated test.ino

uses ; instead of . for string parsing, allowing support for decimal numbers

anaymehta committed:
change testScript.py delimiter to ;

anaymehta committed:
fixes

anaymehta committed:
formatting + ; for interpreter

anaymehta commented on daq/panda/test/interpreter.py:
```diff
@@ -0,0 +1,100 @@
+        for i in range (0, numValues):
+                inputs_dict[prefix[i]] = float(data[i]) #may fail on first run, simply re-run
+        
+        #Problem
```
is there still a problem here? if so, can we get more detail as to what it is in the comment. If not, let's remove this line of code.

anaymehta commented on daq/panda/test/interpreter.py:
```diff
@@ -0,0 +1,100 @@
+text = open(completeName, 'r')
+
+for position, line in enumerate(text):
+    if position%3 == 0:  #reads every third line  
```
why? can we get this line explained by a comment?

anaymehta commented on daq/panda/test/test/readme.md:
```diff
@@ -0,0 +1,5 @@
+main.ino currently prints all received data values in a single string, with each datum spanning a varying number of characters separated by semicolons. This allows main.py to parse the string between semicolons, thereby restoring the original data.
+
+#### Data Formatting
+Each datum is comprised of a label, a colon, and the data value. Ex: Lat:19.4;Lon:92.4;
```
will need to be updated by myself if we decide against labels

anaymehta commented on daq/panda/test/testScript.py:
```diff
@@ -52,17 +58,26 @@
-    data = text[j+4:j+6] #VALUE
-    inputs_dict[prefix] = float(data) #may fail on first run, simply re-run
-    j+=6 #assumes 2 digit data values
+        inputs_dict[prefix[i]] = float(data[i]) #may fail on first run, simply re-run
```
i'm not convinced using a fix prefix array is the best approach, under the assumption that we might have sensor failures wherein we don't receive all 22 expected values when we want them

anaymehta replied:
Being removed!

anaymehta committed:
Update readme.md

Updated readme to reflect new string format

JacobSandler committed:
Merge branch 'daq-string-parsing-documentation' of https://github.com/Bruin-SuperMileage/EV into daq-string-parsing-documentation

JacobSandler committed:
Changed code to be more flexible for input order

JacobSandler committed:
Updated Arduino Test Script for Black Box

Updated the test script to feature 3 letter labels for each data element along with the semicolons to match the current format of the black box.
